src/lib/device/journal_vdev.cpp

```diff
@@ -559,54 +574,108 @@ void JournalVirtualDev::Descriptor::truncate(off_t truncate_offset) {
     }
 
     // Find the chunk which has the truncation offset. This will be the new
-    // head chunk in the list. We first update the is_head is true of this chunk.
+    // head chunk in the list. We first update the is_head to true for that new head chunk.
     // So if a crash happens after this, we could have two chunks which has is_head
     // true in the list and during recovery we select head with the highest creation
-    // timestamp and reuse or cleanup the other.
-    auto [new_head_chunk, _, offset_in_chunk] = offset_to_chunk(truncate_offset);
+    // timestamp and cleanup the other. We check if the truncation offset happens to be
+    // between the end_of_chunk mark and actual chunk end. In that case there is not data
+    // in that chunk and we release this chunk and select the next chunk.
+    bool update_truncate_offset = false;
+    auto [new_head_chunk, index, offset_in_chunk] = offset_to_chunk(truncate_offset);
+    auto end_new_head_chunk = m_vdev.get_end_of_chunk(new_head_chunk);
+    if (offset_in_chunk >= static_cast< int64_t >(end_new_head_chunk)) {
+        // If the truncation offset is same as end_of_chunk, we dont have any more
+        // data in this chunk to read. Select the next chunk.
+        if (++index <= (m_journal_chunks.size() - 1)) {
```

Comment:
what will be consequence for the `else` part of this if?

Reply:
in else branch, we didn't set update_truncate_offset to true, what will be the truncate_offset in else branch? it looks like we rely on `update_truncate_offset` to update data start offset as return value, is it expected the we release all the chunks without updating data start offset?

Reply:
update_data_start_offset is used in this to move to the next chunk size align. We do this when truncate happens to fall in area between end_of_chunk - chunk size or all chunks are truncated. In both cases we roundup to the chunk size. Also update_data_start_offset is in memory, we return the latest data_start_offset back to the caller, which stores in the log dev entry.

Reply:
the question here is in else branch, shouldn't we also update the data start offset since all the chunks are truncated? Without setting update_truncate_offset to true, it will just return what ever is there as a staled value in data_start_offset and return to caller? This part I am not able to get it.

Reply:
Discussed with Yaming.